hutch_python/utils.py

```diff
@@ -433,3 +436,77 @@ def get_fully_qualified_domain_name():
         )
         logger.debug("getfqdn failed", exc_info=True)
         return ""
+
+
+abort_msg = """
+Your RunEngine has been interrupted.  If you aborted the current run (Ctrl+C),
+the RunEngine is ready for use.  If you paused the run (Ctrl+\\), you must
+resume, abort, stop, halt the RunEngine before continuing.
+"""
+
+
+class AbortSigintHandler(SignalHandler):
+    """
+    Interprets SIGINT (Ctrl+C) as an abort request.
+    """
+    def __init__(self, RE):
+        super().__init__(signal.SIGINT, log=RE.log)
+        self.RE = RE
+
+    def handle_signals(self):
+        # Check for pause requests from keyboard.
+        if self.RE.state.is_running:
+            self.RE.stop()
+            # Use our own logging
+            logger.warning("Aborting current run.")
+
+
+class SigquitHandler(SignalHandler):
+    """
+    Interprets SIGQUIT (Ctrl+\\) as standard RunEngine interrupt behavior.
```

Comment:
This seems reasonable from a user/usage perspective, but confusing from a linux perspective. TIL that `SIGQUIT` is the same as `SIGINT` except that the process is supposed to produce a core dump. On balance I'd say this is a good approach but I'm a bit wary.

Reply:
`SIGQUIT` seems to be caught by IPython and exits _without_ a core file, as far as I can tell.

If the concern is strong enough, we could look into what it'd take to hook all keyboard events to filter for the original proposed `Ctrl-P`. 

I guess it'd be good to fully figure out the level of concern prior to going forward, as we wouldn't want to switch out the key combination from under the users down the line.

Reply:
My concern isn't that high of any practical problems arising from this. This seems like the most efficient way to get to the goal of "some keypress to retain the pausing behavior while letting ctrl+c feel like stopping the run entirely". It just seems semantically discordant.

As a side note, understanding good approaches for responding to "any" keypress during something like a scan might be a good thing for us to learn.

Reply:
I guess one concern is that `SIGQUIT` will exit ipython if the RunEngine is not going through the pause routine.  I can see that causing grief in the future if users hit `Ctrl-\` three times and have to re-load hutch-python.  `SIGINT` was really the signal for the job.  

You could argue that our users should be smashing `Ctrl-\` if `Ctrl-c` already aborts, but that might be a risky assumption

Reply:
I guess I was too focused on how nicely this worked in testing (and didn't mash the keys as you might expect users to). I think @tangkong 's point is strong enough to warrant finding an alternative signal - or better, a keyboard hook.

This was also enough to make me dig into the details of keybindings, and it may be easier than I had assumed: https://github.com/ipython/ipython/blob/3587f5bb6c8570e7bbb06cf5f7e3bc9b9467355a/IPython/terminal/shortcuts.py#L58

And even possible without an additional library...

Reply:
Turns out this doesn't work, as the RunEngine thread prevents input from reaching the ipython prompt_toolkit.  

A polling thread watching for a separate key combination has been suggested by Zach multiple times.  I'll look into it a bit.